### ISSUE TITLE
Tests timing out in `wait_for_cql_and_get_hosts`, due to driver failing to reconnect, driver throwing `KeyError` on reconnection attempts

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#17662]
Seen in https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7145/artifact/testlog/x86_64/debug/topology_custom.test_raft_recovery_stuck.2.log
With scylla 5.5.0~dev-0.20240305.6c83167db61d
```
12:05:42.655 DEBUG> [control connection] Refreshing node list and token map using preloaded results
12:05:42.655 DEBUG> [control connection] Found new host to connect to: 127.180.207.41:9042
12:05:42.655 DEBUG> Closing connection (139695759006352) to 127.180.207.25:9042
12:05:42.655 DEBUG> Closed socket to 127.180.207.25:9042
12:05:42.655 WARNING> [control connection] Error connecting to 127.180.207.25:9042:
Traceback (most recent call last):
  File "cassandra/cluster.py", line 3629, in cassandra.cluster.ControlConnection._connect_host_in_lbp
  File "cassandra/cluster.py", line 3737, in cassandra.cluster.ControlConnection._try_connect
  File "cassandra/cluster.py", line 3733, in cassandra.cluster.ControlConnection._try_connect
  File "cassandra/cluster.py", line 3986, in cassandra.cluster.ControlConnection._refresh_node_list_and_token_map
  File "cassandra/cluster.py", line 2172, in cassandra.cluster.Cluster.add_host
  File "cassandra/cluster.py", line 2174, in cassandra.cluster.Cluster.add_host
KeyError: UUID('6631e619-983c-4501-a512-1e813f135f63')
12:05:42.655 DEBUG> [control connection] Opening new connection to 127.180.207.41:9042
12:05:42.655 DEBUG> Connection 139695759012496: '('127.0.0.1', 49418)' -> '('127.180.207.41', 9042)'
12:05:42.655 DEBUG> Sending initial options message for new connection (139695759012496) to 127.180.207.41:9042
12:05:42.656 DEBUG> Parsing sharding info from message options {'COMPRESSION': ['lz4', 'snappy'], 'SCYLLA_LWT_ADD_METADATA_MARK': ['LWT_OPTIMIZATION_META_BIT_MASK=2147483648'], 'SCYLLA_NR_SHARDS': ['2'], 'SCYLLA_PARTITIONER': ['org.apache.cassandra.dht.Murmur3Partitioner'], 'SCYLLA_RATE_LIMIT_ERROR': ['ERROR_CODE=61440'], 'SCYLLA_SHARD': ['1'], 'SCYLLA_SHARDING_ALGORITHM': ['biased-token-round-robin'], 'SCYLLA_SHARDING_IGNORE_MSB': ['12'], 'TABLETS_ROUTING_V1': ['']}
12:05:42.656 DEBUG> Received options response on new connection (139695759012496) from 127.180.207.41:9042
12:05:42.656 DEBUG> No available compression types supported on both ends. locally supported: odict_keys([]). remotely supported: ['lz4', 'snappy']
12:05:42.656 DEBUG> Sending StartupMessage on <CustomConnection(139695759012496) 127.180.207.41:9042>
12:05:42.656 DEBUG> Sent StartupMessage on <CustomConnection(139695759012496) 127.180.207.41:9042>
12:05:42.656 DEBUG> Got ReadyMessage on new connection (139695759012496) from 127.180.207.41:9042
12:05:42.656 DEBUG> [control connection] Established new connection <CustomConnection(139695759012496) 127.180.207.41:9042>, registering watchers and refreshing schema and topology
12:05:42.658 DEBUG> [control connection] Refreshing node list and token map using preloaded results
12:05:42.658 DEBUG> [control connection] Finished fetching ring info
12:05:42.658 DEBUG> Closing connection (139695759012496) to 127.180.207.41:9042
12:05:42.658 DEBUG> Closed socket to 127.180.207.41:9042
12:05:42.658 WARNING> [control connection] Error connecting to 127.180.207.41:9042:
Traceback (most recent call last):
  File "cassandra/cluster.py", line 3629, in cassandra.cluster.ControlConnection._connect_host_in_lbp
  File "cassandra/cluster.py", line 3737, in cassandra.cluster.ControlConnection._try_connect
  File "cassandra/cluster.py", line 3734, in cassandra.cluster.ControlConnection._try_connect
  File "cassandra/cluster.py", line 3836, in cassandra.cluster.ControlConnection._refresh_schema
  File "cassandra/metadata.py", line 141, in cassandra.metadata.Metadata.refresh
AttributeError: 'NoneType' object has no attribute 'release_version'
12:05:42.658 DEBUG> [control connection] Opening new connection to 127.180.207.38:9042
12:05:42.658 DEBUG> Connection 139695664140944: '('127.0.0.1', 38514)' -> '('127.180.207.38', 9042)'
12:05:42.658 DEBUG> Sending initial options message for new connection (139695664140944) to 127.180.207.38:9042
12:05:42.659 DEBUG> Parsing sharding info from message options {'COMPRESSION': ['lz4', 'snappy'], 'SCYLLA_LWT_ADD_METADATA_MARK': ['LWT_OPTIMIZATION_META_BIT_MASK=2147483648'], 'SCYLLA_NR_SHARDS': ['2'], 'SCYLLA_PARTITIONER': ['org.apache.cassandra.dht.Murmur3Partitioner'], 'SCYLLA_RATE_LIMIT_ERROR': ['ERROR_CODE=61440'], 'SCYLLA_SHARD': ['0'], 'SCYLLA_SHARDING_ALGORITHM': ['biased-token-round-robin'], 'SCYLLA_SHARDING_IGNORE_MSB': ['12'], 'TABLETS_ROUTING_V1': ['']}
12:05:42.659 DEBUG> Received options response on new connection (139695664140944) from 127.180.207.38:9042
12:05:42.659 DEBUG> No available compression types supported on both ends. locally supported: odict_keys([]). remotely supported: ['lz4', 'snappy']
12:05:42.659 DEBUG> Sending StartupMessage on <CustomConnection(139695664140944) 127.180.207.38:9042>
12:05:42.659 DEBUG> Sent StartupMessage on <CustomConnection(139695664140944) 127.180.207.38:9042>
12:05:42.659 DEBUG> Got ReadyMessage on new connection (139695664140944) from 127.180.207.38:9042
12:05:42.659 DEBUG> [control connection] Established new connection <CustomConnection(139695664140944) 127.180.207.38:9042>, registering watchers and refreshing schema and topology
12:05:42.661 DEBUG> [control connection] Refreshing node list and token map using preloaded results
12:05:42.661 DEBUG> [control connection] Found new host to connect to: 127.180.207.41:9042
12:05:42.661 DEBUG> Closing connection (139695664140944) to 127.180.207.38:9042
12:05:42.661 DEBUG> Closed socket to 127.180.207.38:9042
12:05:42.661 WARNING> [control connection] Error connecting to 127.180.207.38:9042:
Traceback (most recent call last):
  File "cassandra/cluster.py", line 3629, in cassandra.cluster.ControlConnection._connect_host_in_lbp
  File "cassandra/cluster.py", line 3737, in cassandra.cluster.ControlConnection._try_connect
  File "cassandra/cluster.py", line 3733, in cassandra.cluster.ControlConnection._try_connect
  File "cassandra/cluster.py", line 3986, in cassandra.cluster.ControlConnection._refresh_node_list_and_token_map
  File "cassandra/cluster.py", line 2172, in cassandra.cluster.Cluster.add_host
  File "cassandra/cluster.py", line 2174, in cassandra.cluster.Cluster.add_host
KeyError: UUID('6631e619-983c-4501-a512-1e813f135f63')
12:05:42.661 DEBUG> [control connection] error reconnecting
Traceback (most recent call last):
  File "cassandra/cluster.py", line 3750, in cassandra.cluster.ControlConnection._reconnect
  File "cassandra/cluster.py", line 3658, in cassandra.cluster.ControlConnection._reconnect_internal
  File "cassandra/cluster.py", line 1738, in cassandra.cluster.Cluster._add_resolved_hosts
  File "cassandra/cluster.py", line 2172, in cassandra.cluster.Cluster.add_host
  File "cassandra/cluster.py", line 2174, in cassandra.cluster.Cluster.add_host
KeyError: UUID('6631e619-983c-4501-a512-1e813f135f63')
12:05:43.430 DEBUG> Sending options message heartbeat on idle connection (139695764862608) 127.180.207.38:9042
12:05:43.430 DEBUG> Sending options message heartbeat on idle connection (139695763925968) 127.180.207.38:9042
12:05:43.431 DEBUG> Sending options message heartbeat on idle connection (139695763924112) 127.180.207.25:9042
12:05:43.431 DEBUG> Received options response on connection (139695764862608) from 127.180.207.38:9042
12:05:43.431 DEBUG> Received options response on connection (139695763925968) from 127.180.207.38:9042
12:05:43.431 DEBUG> Received options response on connection (139695763924112) from 127.180.207.25:9042
12:05:43.652 DEBUG> [control connection] Refreshing node list and token map
12:05:43.655 DEBUG> [control connection] Found new host to connect to: 127.180.207.41:9042
12:05:43.655 DEBUG> [control connection] Error refreshing node list and token map
Traceback (most recent call last):
  File "cassandra/cluster.py", line 3843, in cassandra.cluster.ControlConnection.refresh_node_list_and_token_map
  File "cassandra/cluster.py", line 3986, in cassandra.cluster.ControlConnection._refresh_node_list_and_token_map
  File "cassandra/cluster.py", line 2172, in cassandra.cluster.Cluster.add_host
  File "cassandra/cluster.py", line 2174, in cassandra.cluster.Cluster.add_host
KeyError: UUID('6631e619-983c-4501-a512-1e813f135f63')
---------------------------- Captured log teardown -----------------------------
12:05:43.681 DEBUG> after_test for test_recover_stuck_raft_recovery (success: False)
12:05:43.681 DEBUG> RESTClient fetching PUT http+unix://api/cluster/after-test/False
12:05:43.682 INFO> Cluster after test test_recover_stuck_raft_recovery: ScyllaCluster(name: b94ad25e-dad7-11ee-ba36-a8a159ead535, running: ScyllaServer(596, 127.180.207.25, 09c78a66-b40b-4b0c-8494-42537cbd8ade), ScyllaServer(604, 127.180.207.41, 60fe7569-1501-46b8-8d52-6e9f3908d548), ScyllaServer(597, 127.180.207.38, f45d7ae7-fd05-4707-9035-d774c4fe0a5c), stopped: )
```

**Comment by bhalevy** [comment]
@kbr-scylla can you please assign someone to look into this?
Could be pylib issue or driver issue maybe?

**Comment by bhalevy** [comment]
Seen also in https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7131/artifact/testlog/x86_64/debug/topology_custom.test_raft_recovery_basic.3.log

**Comment by kbr-scylla** [comment]
Looks like the KeyError is thrown from new code in python driver (scylladb/python-driver@f2a80f1772e8addb6ea6bc5171c63b8854ef0aa7)

@avelanarius @sylwiaszunejko

**Comment by sylwiaszunejko** [comment]
@kbr-scylla I will look at it

**Comment by kbr-scylla** [comment]
Generalized title. We're observing this in multiple tests.

**Comment by kbr-scylla** [comment]
Another occurrence: https://github.com/scylladb/scylladb/issues/17672

**Comment by sylwiaszunejko** [comment]
@kbr-scylla I am trying to reproduce it locally, but it does not reproduce. Do these failures have some specific settings I should use?

**Comment by kbr-scylla** [comment]
> Do these failures have some specific settings I should use?

Nothing I'm aware of.

It's a concurrency bug (like 99% bugs in the Python driver). Some steps you could try that helped me in the past when I tried to reproduce concurrency issues:
- run many concurrent tests. You can increase the pool size used by test.py suites, for example if you're running a test in test/topology_custom suite, modify test/topology_custom/suite.yaml and increase pool_size to a larger number, then if you run with `--repeat X` (where X is larger than pool_size) you'll have that many concurrent tests running in this suite
- limit the available CPU cores so concurrent tests have to compete for resources more, by passing `--cpus 0,1` to test.py (this example limits to cores 0 and 1)
- often when I cannot reproduce a concurrency issue, I try to come up with a hypothesis and then orchestrate the failure by injecting sleeps tactically in the code based on what I think could be happening.
- you can also try running on CI using this job https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/

And of course make sure you're using the latest python driver version, it must be a regression in 3.26.7.

**Comment by kbr-scylla** [comment]
@sylwiaszunejko you could also have more luck reproducing on Armatrading, this machine likes concurrency bugs.
Ask @yaronkaikov for access if you don't have already

**Comment by sylwiaszunejko** [comment]
@yaronkaikov I do not have access to Armatrading, could you help me with that?

**Comment by sylwiaszunejko** [comment]
It took many runs, but I managed to reproduce locally with `topology_custom.test_replace_ignore_nodes`

**Comment by benipeled** [comment]
Seen on master gating https://jenkins.scylladb.com/job/scylla-master/job/next/7365/artifact/testlog/x86_64/debug/topology_custom.test_raft_recovery_basic.2.log/*view*/

**Comment by benipeled** [comment]
...and enterprise https://jenkins.scylladb.com/job/scylla-enterprise/job/next/2844/artifact/testlog/x86_64/debug/topology_custom.test_raft_recovery_stuck.2.log/*view*/

**Comment by denesb** [comment]
This is failing CI and promotion multiple times, every day.

**Comment by avikivity** [comment]
If Python driver concurrency bugs are hurting CI, it means the Python driver needs a concurrency test suite independent from scylladb.git.

**Comment by Annamikhlin** [comment]
seen on 
https://jenkins.scylladb.com/job/scylla-master/job/next/7392/testReport/junit/(root)/test_raft_recovery_basic/Build___aarch64___Unit_Tests_aarch64___test_raft_recovery_basic_2/

**Comment by bhalevy** [comment]
I hit this locally also with test/topology/test_topology_remove_decom.py test_rebuild_node

**Comment by Annamikhlin** [comment]
`test_replace_ignore_nodes` test also failed on timeout in `wait_for_cql_and_get_hosts`
https://jenkins.scylladb.com/job/scylla-master/job/next/7400/testReport/junit/(root)/test_replace_ignore_nodes/Build___aarch64___Unit_Tests_aarch64___test_replace_ignore_nodes/

link to log: https://jenkins.scylladb.com/job/scylla-master/job/next/7400/artifact/testlog/aarch64/dev/topology_custom.test_replace_ignore_nodes.1.log

**Comment by Annamikhlin** [comment]
seen also here:
https://jenkins.scylladb.com/job/scylla-master/job/next/7409/testReport/junit/(root)/test_raft_recovery_stuck/Build___x86___Unit_Tests_x86___test_recover_stuck_raft_recovery/

**Comment by sylwiaszunejko** [comment]
@avikivity https://github.com/scylladb/python-driver/pull/308 PR with the fix was merged and new driver version was released (3.26.8).
